waiter/src/waiter/auth/authentication.clj

```diff
@@ -26,12 +27,26 @@
   (wrap-auth-handler [this request-handler]
     "Attaches middleware that enables the application to perform authentication.
      The middleware should
-     - either issue a 401 challenge asking the client to authenticate itself,
-     - or upon successful authentication populate the request with :authorization/user and :authorization/principal"))
+     - issue a 401 challenge, or redirect, to get the client to authenticate itself,
+     - or upon successful authentication populate the request with :authorization/user and :authorization/principal")
+  (get-authentication-providers [this]
+    "Get a list of supported authenticaiton provider names.")
+  (process-callback [this request]
+    "Process any requests that might come in after initiating authentication. e.g. receive a request
+     with an authentication assertion after redirecting a user to authenticate with an identity provider."))
+
+(extend-protocol Authenticator
```

Comment:
This is a nice way to introduce backward compatible changes when we introduce new methods on the protocols.

Should we still go ahead and modify the existing authenticator implementations in this codebase?

> To create a new protocol that operates on an existing datatype and when you cannot modify the source code of the defrecord.

https://www.codementor.io/mounacheikhna/simple-clojure-protocols-tutorial-ss8pywzkd#extending-protocols-to-already-existing-types

Reply:
I like the "default" methods in one place.

Reply:
We should see how others in the team feel for resolution :)

Reply:
We can add a docstring in the Authenticator to say standard is returned as the default and then we don't need to modify the existing authenticators.

Reply:
see other discussion